File: hrms/hr/report/employee_advance_summary/employee_advance_summary.js
// Copyright (c) 2016, Frappe Technologies Pvt. Ltd. and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Employee Advance Summary"] = {
	filters: [
		{
			fieldname: "employee",
			label: __("Employee"),
			fieldtype: "Link",
			options: "Employee",
			width: "80",
		},
		{
			fieldname: "from_date",
			label: __("From Date"),
			fieldtype: "Date",
			default: erpnext.utils.get_fiscal_year(frappe.datetime.get_today(), true)[1],
			width: "80",
		},
		{
			fieldname: "to_date",
			label: __("To Date"),
			fieldtype: "Date",
			default: frappe.datetime.get_today(),
		},
		{
			fieldname: "company",
			label: __("Company"),
			fieldtype: "Link",
			options: "Company",
			default: frappe.defaults.get_user_default("Company"),
			reqd: 1,
		},
		{
			fieldname: "status",
			label: __("Status"),
			fieldtype: "Select",
			options: "\nDraft\nPaid\nPartially Paid\nUnpaid\nClaimed\nCancelled",
		},
		{
			fieldname: "department",
			label: __("Department"),
			fieldtype: "MultiSelectList",
			options: "Department",
			get_data: function (txt) {
				return frappe.db.get_link_options("Department", txt);
			},
		},
		{
			fieldname: "branch",
			label: __("Branch"),
			fieldtype: "MultiSelectList",
			options: "Branch",
			get_data: function (txt) {
				return frappe.db.get_link_options("Branch", txt);
			},
		},
		{
			fieldname: "advance_account",
			label: __("Advance Account"),
			fieldtype: "MultiSelectList",
			options: "Account",
			get_data: function (txt) {
				var company = frappe.query_report.get_filter_value("company");
				return frappe.db.get_link_options("Account", txt, {
					company: company,
					account_type: "Receivable",
				});
			},
		},
		{
			fieldname: "group_by",
			label: __("Group By"),
			fieldtype: "Select",
			options: "\nEmployee\nDepartment\nBranch",
		},
	],

	formatter: function (value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		if (data && data.bold) {
			value = `<strong>${value}</strong>`;
		}
		return value;
	},
};

File: hr_app/hr/report/employee_advance_summary/employee_advance_summary.py
# Copyright (c) 2013, Frappe Technologies Pvt. Ltd. and contributors
# For license information, please see license.txt


from pypika import Order

import frappe
from frappe import _, msgprint


def execute(filters=None):
	if not filters:
		filters = {}

	advances_list = get_advances(filters)
	columns = get_columns()

	if not advances_list:
		msgprint(_("No record found"))
		return columns, advances_list

	data = []
	for advance in advances_list:
		row = [
			advance.name,
			advance.employee,
			advance.company,
			advance.posting_date,
			advance.advance_amount,
			advance.paid_amount,
			advance.claimed_amount,
			advance.return_amount,
			advance.status,
			advance.currency,
		]
		data.append(row)

	return columns, data


def get_columns():
	return [
		{
			"label": _("Title"),
			"fieldname": "title",
			"fieldtype": "Link",
			"options": "Employee Advance",
			"width": 120,
		},
		{
			"label": _("Employee"),
			"fieldname": "employee",
			"fieldtype": "Link",
			"options": "Employee",
			"width": 120,
		},
		{
			"label": _("Company"),
			"fieldname": "company",
			"fieldtype": "Link",
			"options": "Company",
			"width": 120,
		},
		{"label": _("Posting Date"), "fieldname": "posting_date", "fieldtype": "Date", "width": 120},
		{
			"label": _("Advance Amount"),
			"fieldname": "advance_amount",
			"fieldtype": "Currency",
			"options": "currency",
			"width": 120,
		},
		{
			"label": _("Paid Amount"),
			"fieldname": "paid_amount",
			"fieldtype": "Currency",
			"options": "currency",
			"width": 120,
		},
		{
			"label": _("Claimed Amount"),
			"fieldname": "claimed_amount",
			"fieldtype": "Currency",
			"options": "currency",
			"width": 120,
		},
		{
			"label": _("Returned Amount"),
			"fieldname": "return_amount",
			"fieldtype": "Currency",
			"options": "currency",
			"width": 120,
		},
		{"label": _("Status"), "fieldname": "status", "fieldtype": "Data", "width": 120},
		{
			"label": _("Currency"),
			"fieldtype": "Link",
			"fieldname": "currency",
			"options": "Currency",
			"hidden": 1,
			"width": 120,
		},
	]


def get_advances(filters):
	EmployeeAdvance = frappe.qb.DocType("Employee Advance")

	query = (
		frappe.qb.from_(EmployeeAdvance)
		.select(
			EmployeeAdvance.name,
			EmployeeAdvance.employee,
			EmployeeAdvance.paid_amount,
			EmployeeAdvance.status,
			EmployeeAdvance.advance_amount,
			EmployeeAdvance.claimed_amount,
			EmployeeAdvance.return_amount,
			EmployeeAdvance.company,
			EmployeeAdvance.posting_date,
			EmployeeAdvance.purpose,
			EmployeeAdvance.currency,
		)
		.where(EmployeeAdvance.docstatus < 2)
	)

	if filters.get("employee"):
		query = query.where(EmployeeAdvance.employee == filters.employee)

	if filters.get("company"):
		query = query.where(EmployeeAdvance.company == filters.company)

	if filters.get("status"):
		query = query.where(EmployeeAdvance.status == filters.status)

	if filters.get("from_date"):
		query = query.where(EmployeeAdvance.posting_date >= filters.from_date)

	if filters.get("to_date"):
		query = query.where(EmployeeAdvance.posting_date <= filters.to_date)

	return query.orderby(EmployeeAdvance.posting_date, EmployeeAdvance.name, order=Order.desc).run(
		as_dict=True
	)
